File: pc4/JsWorker/handlers/uploadHandler.js
const fs = require('fs-extra');
const path = require('path');
const axios = require('axios');
const FormData = require('form-data');
const RaftNode = require('../raftNode'); // Asegúrate de que la ruta es correcta

const STORAGE_DIR = path.join(__dirname, '../storage');

const workerAddresses = ["http://pyworker:8080", "http://jsworker:8081", "http://javaworker:8082"];
const raftNode = new RaftNode("jsworker", workerAddresses);
raftNode.run();

const replicateFile = (filePath, fileName) => {
    workerAddresses.forEach(worker => {
        if (worker !== `http://jsworker:8081`) {
            const formData = new FormData();
            formData.append('file', fs.createReadStream(filePath));
            axios.post(`${worker}/upload`, formData, {
                headers: formData.getHeaders()
            }).then(response => {
                console.log(`Replicating ${fileName} to ${worker}: ${response.status}`);
            }).catch(error => {
                console.error(`Error replicating ${fileName} to ${worker}: ${error.message}`);
            });
        }
    });
};

const handleUpload = (req, res) => {
    console.log('Handling file upload...');
    console.log('Request headers:', req.headers);

    const contentType = req.headers['content-type'];
    if (!contentType || !contentType.startsWith('multipart/form-data')) {
        return res.status(400).json({ error: 'Content-Type must be multipart/form-data' });
    }

    let fileBuffer = Buffer.alloc(0);
    req.on('data', (chunk) => {
        fileBuffer = Buffer.concat([fileBuffer, chunk]);
    });

    req.on('end', () => {
        console.log('Received file data');
        const boundary = '--' + contentType.split('boundary=')[1];
        const parts = fileBuffer.toString().split(boundary).slice(1, -1);

        if (parts.length === 0) {
            return res.status(400).json({ error: 'No files were uploaded.' });
        }

        const filePart = parts[0];
        const [header, fileContent] = filePart.split('\r\n\r\n');
        const headerLines = header.toString().split('\r\n');
        const filenameLine = headerLines.find(line => line.includes('filename='));
        const filename = filenameLine.match(/filename="([^"]+)"/)[1].trim();

        console.log('Saving file:', filename);
        const filePath = path.join(STORAGE_DIR, filename);

        // Write the file content to the file system
        fs.writeFile(filePath, fileContent.trim(), (err) => {
            if (err) {
                console.error('Error saving file:', err.message);
                return res.status(500).json({ error: err.message });
            }
            if (raftNode.getState() === 'leader') {
                replicateFile(filePath, filename);
            }
            res.json({ message: 'File uploaded successfully' });
        });
    });

    req.on('error', (err) => {
        console.error('Request error:', err.message);
        res.status(500).json({ error: err.message });
    });
};

module.exports = { handleUpload };

File: pc4/JavaWorker/src/main/java/com/worker/UploadHandler.java
package com.worker;

import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpExchange;
import com.google.gson.Gson;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class UploadHandler implements HttpHandler {
    private static final Gson gson = new Gson();
    private static final String[] WORKERS = {
        "http://pyworker:8080",
        "http://jsworker:8081",
        "http://javaworker:8082"
    };
    private final RaftNode raftNode;

    public UploadHandler(RaftNode raftNode) {
        this.raftNode = raftNode;
    }

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        System.out.println("Handling file upload...");

        if (!"POST".equals(exchange.getRequestMethod())) {
            System.out.println("Invalid request method: " + exchange.getRequestMethod());
            exchange.sendResponseHeaders(405, -1); // Method Not Allowed
            return;
        }

        String contentType = exchange.getRequestHeaders().getFirst("Content-Type");
        if (contentType == null || !contentType.startsWith("multipart/form-data")) {
            System.out.println("Invalid content type: " + contentType);
            exchange.sendResponseHeaders(400, -1); // Bad Request
            return;
        }

        String boundary = "--" + contentType.split("boundary=")[1];
        InputStream inputStream = exchange.getRequestBody();
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        byte[] chunk = new byte[1024];
        int bytesRead;

        while ((bytesRead = inputStream.read(chunk)) != -1) {
            buffer.write(chunk, 0, bytesRead);
        }

        byte[] body = buffer.toByteArray();
        String bodyString = new String(body);
        String[] parts = bodyString.split(boundary);

        for (String part : parts) {
            if (part.contains("filename=\"")) {
                System.out.println("Processing part: " + part);
                String[] headers = part.split("\r\n");
                String fileName = headers[1].split("filename=\"")[1].split("\"")[0];
                int fileStart = part.indexOf("\r\n\r\n") + 4;
                int fileEnd = part.lastIndexOf("\r\n--");
                if (fileEnd == -1) {
                    fileEnd = part.length() - 2;
                }
                byte[] fileContent = part.substring(fileStart, fileEnd).getBytes();

                Files.write(Paths.get("storage/" + fileName), fileContent);
                System.out.println("File " + fileName + " uploaded successfully");
                if (raftNode.getState().equals("leader")) {
                    replicateFile("storage/" + fileName, fileName);
                }
                Map<String, String> response = new HashMap<>();
                response.put("message", "File uploaded successfully");
                String jsonResponse = gson.toJson(response);
                exchange.getResponseHeaders().set("Content-Type", "application/json");
                exchange.sendResponseHeaders(200, jsonResponse.length());
                OutputStream os = exchange.getResponseBody();
                os.write(jsonResponse.getBytes());
                os.close();
                return;
            }
        }

        System.out.println("No files were uploaded.");
        exchange.sendResponseHeaders(400, -1); // Bad Request
    }

    private void replicateFile(String filePath, String fileName) {
        for (String worker : WORKERS) {
            if (!worker.equals("http://javaworker:8082")) {
                try {
                    File file = new File(filePath);
                    HttpURLConnection connection = (HttpURLConnection) new URL(worker + "/upload").openConnection();
                    connection.setDoOutput(true);
                    connection.setRequestMethod("POST");
                    connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=---ContentBoundary");

                    OutputStream os = connection.getOutputStream();
                    PrintWriter writer = new PrintWriter(new OutputStreamWriter(os), true);

                    writer.append("---ContentBoundary\r\n");
                    writer.append("Content-Disposition: form-data; name=\"file\"; filename=\"" + fileName + "\"\r\n");
                    writer.append("Content-Type: " + Files.probeContentType(file.toPath()) + "\r\n\r\n");
                    writer.flush();

                    Files.copy(file.toPath(), os);
                    os.flush();

                    writer.append("\r\n---ContentBoundary--\r\n");
                    writer.flush();

                    int responseCode = connection.getResponseCode();
                    System.out.println("Replicating " + fileName + " to " + worker + ": " + responseCode);

                    writer.close();
                    os.close();
                } catch (Exception e) {
                    System.err.println("Error replicating " + fileName + " to " + worker + ": " + e.getMessage());
                }
            }
        }
    }
}
